test/e2e/storage/csi_volumes.go

```diff
@@ -367,57 +369,96 @@ func initCSIgcePD(f *framework.Framework, config framework.VolumeTestConfig) csi
+		patchCSIDeployment(g.f, "com.google.csi.gcepd", "gce-driver", "csi-external-provisioner", g.config.ServerNodeName, object)
+	},
+		"test/e2e/testing-manifests/storage-csi/driver-registrar/rbac.yaml",
+		"test/e2e/testing-manifests/storage-csi/external-attacher/rbac.yaml",
+		"test/e2e/testing-manifests/storage-csi/external-provisioner/rbac.yaml",
+		"test/e2e/testing-manifests/storage-csi/gce-pd/csi-controller-rbac.yaml",
+		"test/e2e/testing-manifests/storage-csi/gce-pd/node_ds.yaml",
+		"test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss.yaml",
+		"test/e2e/testing-manifests/storage-csi/gce-pd/controller_service.yaml",
+	)
+	g.cleanup = cleanup
+	if err != nil {
+		framework.Failf("deploying csi hostpath driver: %v", err)
+	}
 }
 
 func (g *gcePDCSIDriver) cleanupCSIDriver() {
 	By("uninstalling gce-pd driver")
-	f := g.f
-	cs := f.ClientSet
-	config := g.config
-	deployGCEPDCSIDriver(cs, config, true /* teardown */, f, g.nodeServiceAccount, g.controllerServiceAccount)
-	csiClusterRoleBindings(cs, config, true /* teardown */, g.controllerServiceAccount, g.controllerClusterRoles)
-	csiClusterRoleBindings(cs, config, true /* teardown */, g.nodeServiceAccount, g.nodeClusterRoles)
-	utils.PrivilegedTestPSPClusterRoleBinding(cs, config.Namespace, true, /* teardown */
-		[]string{g.controllerServiceAccount.Name, g.nodeServiceAccount.Name})
-	role := csiControllerRole(cs, config, true)
-	csiControllerRoleBinding(cs, config, true, role, g.controllerServiceAccount)
-	csiServiceAccount(cs, config, "gce-controller", true /* teardown */)
-	csiServiceAccount(cs, config, "gce-node", true /* teardown */)
+	if g.cleanup != nil {
+		g.cleanup()
+	}
+}
+
+// TODO (?): the storage.csi.image.version and storage.csi.image.registry
+// settings are ignored. We could patch the image definitions.
+// TODO: way too many parameters...
+func patchCSIDeployment(f *framework.Framework, provisioner, driverContainerName, provisionerContainerName, serverNodeName string, object interface{}) {
+	patchVolumes := func(volumes []v1.Volume) {
+		for i := range volumes {
+			volume := &volumes[i]
+			if volume.HostPath != nil {
+				// Update paths like /var/lib/kubelet/plugins/<provisioner>.
+				path := &volume.HostPath.Path
+				if strings.HasSuffix(*path, "/"+provisioner) {
+					*path = *path + "-" + f.UniqueName
+				}
+			}
+		}
+	}
+
+	patchContainers := func(containers []v1.Container) {
+		for i := range containers {
+			container := &containers[i]
+			for e := range container.Args {
+				// Inject test-specific provider name into paths (--kubelet-registration-path).
+				container.Args[e] = strings.Replace(container.Args[e], "/"+provisioner+"/", "/"+provisioner+"-"+f.UniqueName+"/", 1)
+			}
+			// Overwrite driver name resp. provider name
+			// by appending a parameter with the right
+			// value.
+			switch container.Name {
+			case driverContainerName:
+				container.Args = append(container.Args, "--drivername="+provisioner+"-"+f.UniqueName)
```

Comment:
Hmm, the gcePD test passed, so the driver must have accepted `--drivername`:
```
I1022 21:12:26.383]             "image": "gcr.io/google-containers/volume-csi/gcp-compute-persistent-disk-csi-driver:v0.1.0.alpha",
I1022 21:12:26.383]             "args": [
I1022 21:12:26.383]               "--v=5",
I1022 21:12:26.383]               "--endpoint=$(CSI_ENDPOINT)",
I1022 21:12:26.383]               "--nodeid=$(KUBE_NODE_NAME)",
I1022 21:12:26.384]               "--drivername=com.google.csi.gcepd-e2e-tests-csi-volumes-w9hsx"
I1022 21:12:26.384]             ],
```
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/69868/pull-kubernetes-e2e-gce/56688/build-log.txt

The patch function in `csi_volumes.go` is driver-specific, it just happens that it intentionally was written so that it works for both hostPath and gcePD. If some other driver doesn't accept the `--drivername` parameter, then it has to use its own patch function or nil, in which case the driver name won't get changed.

Let me add some more documentation to that function to clarify that.

Reply:
Done, and I enhanced the configuration of that function and moved it to the utils package. There it will be easier to import into out-of-tree CSI driver test suites.

Reply:
Ah the latest stable version of the PD driver (referenced above) doesn't support it anymore. So it will have to be changed when we update the PD specs

Reply:
Hmm, so the latest PD driver will have to be used without driver renaming? Can it be guaranteed that there won't be conflicts between different tests without sorting out the "install driver only once" problem first?

Reply:
cc @davidz627 can you confirm driver name is no longer configurable?  we may need to add it back to allow for this testing model.

For out of tree GCE PD testing, we plan to launch the driver as part of cluster setup and not during the e2e test.

Reply:
driver name is no longer configurable (https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/129). This was done because we need a stable well-known driver name for migration. Is there no other way to support testing?